wpep0006.md

```diff
@@ -0,0 +1,97 @@
+| **Authors** | Pascal Cadotte Michaud <pcm@wazo.io> |
+| **Status** | proposed |
+
+## Abstract
+
+Use the UUID of the resources instead of a name for Asterisk resources
+
+## Motivation
+
+In a multi-tenant using a user selected name for Asterisk resources does not make sense.
+
+For example, on a given stack with 100 tenants, only one of the tenants get to call its
+greeting queue `greeting`. All others will get an error when creating the resource.
+
+The name of the Asterisk resource should be generated by the system in order to avoid collision
+between tenants.
+
+## Constraints
+
+As an administrator I want to be able to find the resource easily when debugging and I want
+to know the "asterisk" name when looking at the at my API resources.
+
+The max length of an object in an Asterisk configuration file is 64
+
+## Propositions
+
+### Proposition 1
+
+Each tenant has a `slug` that is a short string.
+The `slug` is unique accross all tenants of a stack.
+
+Each resource has a `UUID`. This `UUID` replaces the `ID` in API calls.
+The `ID` field is only preserved when necessary. Users need an `ID` for function keys.  Ex: `*735<user ID>***225` for the DND FK
+
+Each type of resource has a pre-defined prefix.
+
+USR, GRP, QUE, CONF, etc
+
+Each resource the gets an "asterisk" name
+
+`<resource prefix>-<tenant-slug>-<resource UUID>`
+
+For example
+
+a tenant with the slug `abc123`
+
+Could have a group named `grp-abc123-c007e84a-600c-494d-8a95-6a0d9994d733`
+
+The generated configuration files would also include some information to help debuging
+
+```
+; Tenant abc123 bcbc495d-3f74-45c9-8c6a-a6ff2ab39562
+; Label greeting
+[grp-abc123-c007e84a-600c-494d-8a95-6a0d9994d733]
+```
+
+More over, we could also add some information in the Asterisk log file
+
+```
+[Feb  3 14:02:37] Reaching group greeting@bcbc495d-3f74-45c9-8c6a-a6ff2ab39562
+[Feb  3 14:02:39]     -- Executing [s@queue:32] Queue("SCCP/fkwk2z0g-00000003", "grp-abc123-c007e84a-600c-494d-8a95-6a0d9994d733,i,,,60,,,,,,") in new stack
+```
+
+The API could also return the `asterisk_name` field that contains `grp-abc123-c007e84a-600c-494d-8a95-6a0d9994d733` that cannot be changed
+but can be used to debug this given queue.
+
+### Proposition 2
+
+Similar to Proposition 1 with some differences:
+
+Each resource gets a slug:
+
+* User `Alice Nelligan` = anelligan
+* Conference Room  `Daily Scrum` = dailyscrum
+* Group `Super Duper Long Group Name` = superduper
+* Group `Super Duper Duper Long Group Name` = superdupe2
+* ...
+
+Ideally, slugs are unique within a tenant, so that they can be used as identifier in the API but that is harder to do automatically.
+
+Each resource then gets a Asterisk name:
+
+`<resource>-<tenant slug>-<resource slug>-<resource uuid>`
+
+With the following constraints:
+
+* `<resource>` length max 4 characters
+* `<tenant slug>` length max 10 characters
+* `<resource slug>` length max 10 characters
+* `<resource uuid>` length 36 characters
```

Comment:
What would be the `resource_uuid` utility if the slug is already there and unique by tenant?

Reply:
The UUID can be used in an URL? :shrug: I agree that it is redundant

Reply:
Yes, it is redundant. When debugging, you sometimes have the name, sometimes have the UUID. This make the matching easier, mainly with Asterisk logs where the only thing that appears is the Asterisk name. As Pascal said earlier, this can be done also with comments in files and extra debugging info in the dialplan.